gateway: concurrent checkpoint storage races on shared worktree directory ('Directory not empty', 'worktree add --detach' fail)

## Summary

Around 2026-04-25T20:10:50Z, the gateway's checkpoint handler logged a burst of failures storing checkpoints to the `egg-checkpoints` repo. Multiple checkpoints fail nearly simultaneously, with two distinct error signatures:

```
2026-04-25 20:10:50 [ERROR] gateway.checkpoint-handler: Failed to store checkpoint
  error="Command '['git', '-c', 'core.hooksPath=/dev/null', 'worktree', 'add', '--deta..."
  checkpoint_id=ckpt-7a5fab480ae179fc

2026-04-25 20:10:50 [ERROR] gateway.checkpoint-handler: Failed to store checkpoint
  error="[Errno 39] Directory not empty: '/tmp/checkpoint_2qeqdt8t/checkpoints'"
  checkpoint_id=ckpt-ec29427425bfea5e
```

The pattern (5 failures within a 2-second window, all keyed on shared `/tmp/checkpoint_*/checkpoints` paths) suggests concurrent writers racing on the same worktree directory:
- One checkpoint creates `/tmp/checkpoint_<X>` and `git worktree add` succeeds
- A second checkpoint races into the same `/tmp/checkpoint_<X>` (same per-process tmp prefix? same worktree path arg?) and gets `Errno 39` on the directory or fails the `worktree add`
- Subsequent retries then can't `add --detach` because the worktree slot is taken

## Suggested investigation

1. Confirm whether `/tmp/checkpoint_*/checkpoints` is per-checkpoint (should be unique) or shared across concurrent writers — `checkpoint_handler.py:1069` is the relevant call site.
2. If shared by design, serialize via a per-checkpoint-repo lock; if accidental, mint a unique tmp dir per checkpoint (UUID4 suffix).
3. Audit the cleanup path — the `Errno 39` ('Directory not empty') failure pattern suggests a stale worktree or `.git/worktrees/<name>` entry isn't being cleaned up between checkpoint attempts.

## Severity

p2 — checkpointing is best-effort; a missing checkpoint isn't a pipeline blocker. But this happens during normal pipeline operation and floods the logs with errors that obscure real problems. Likely not specific to a single pipeline; the error window correlates with the gateway-side activity around the (separate) idle-session-prune issue I'm filing in parallel.

## Severity adjusters

- If checkpoint failures cause cascading session-state issues (e.g. resume fails because no checkpoint exists), bump to p1.
- The first error in the burst includes `checkpoint_repo=jwbron/egg-checkpoints` so this looks like the existing checkpoint-store flow, not a one-off.

--- a/gateway/checkpoint_handler.py
+++ b/gateway/checkpoint_handler.py
@@ -879,9 +879,15 @@ class CheckpointHandler:
             return False
 
         target = _resolve_checkpoint_target(checkpoint_repo, remote, repo_path)
+        repo_lock = _get_repo_lock(repo_path, target)
 
         try:
-            with tempfile.TemporaryDirectory(prefix="checkpoint_") as temp_dir:
+            with (
+                repo_lock,
+                tempfile.TemporaryDirectory(
+                    prefix="checkpoint_", ignore_cleanup_errors=True
+                ) as temp_dir,
+            ):
                 temp_path = Path(temp_dir)
 
                 branch_exists = self._branch_exists(
@@ -1062,6 +1068,13 @@ class CheckpointHandler:
                     self._run_git(
                         repo_path,
                         ["worktree", "remove", "--force", str(temp_path)],
+                        check=False,
+                    )
+                    # Drop any stale .git/worktrees/<basename> entry so a
+                    # failed remove doesn't leak across attempts.
+                    self._run_git(
+                        repo_path,
+                        ["worktree", "prune"],
                         check=False,
                     )
 
@@ -1299,6 +1312,23 @@ class CheckpointHandler:
 # Global checkpoint handler instance
 _checkpoint_handler: CheckpointHandler | None = None
 _handler_lock = threading.Lock()
+
+# Per-(repo_path, target) locks serializing concurrent checkpoint stores.
+# Concurrent threads operating on the same source repo race on
+# .git/worktrees state and ref locks, producing 'worktree add' failures
+# and stale worktree directories. See #2069.
+_repo_locks: dict[tuple[str, str], threading.Lock] = {}
+_repo_locks_guard = threading.Lock()
+
+
+def _get_repo_lock(repo_path: str, target: str) -> threading.Lock:
+    key = (repo_path, target)
+    with _repo_locks_guard:
+        lock = _repo_locks.get(key)
+        if lock is None:
+            lock = threading.Lock()
+            _repo_locks[key] = lock
+        return lock
 
 
 def get_checkpoint_handler(github_token: str | None = None) -> CheckpointHandler:

--- a/gateway/checkpoint_handler.py
+++ b/gateway/checkpoint_handler.py
@@ -879,7 +879,7 @@ class CheckpointHandler:
             return False
 
         target = _resolve_checkpoint_target(checkpoint_repo, remote, repo_path)
-        repo_lock = _get_repo_lock(repo_path, target)
+        repo_lock = _get_repo_lock(repo_path)
 
         try:
             with (
@@ -894,79 +894,89 @@ class CheckpointHandler:
                     repo_path, target, CHECKPOINT_BRANCH, github_token=github_token
                 )
 
-                if branch_exists:
-                    # Force-update the local branch to match the remote.
-                    # The + prefix handles the case where the local branch
-                    # has diverged (e.g., from a different remote or
-                    # concurrent checkpoint pushes).
-                    # Retry with backoff: network fetches can be slow or
-                    # transiently fail, especially during container shutdown.
-                    fetch_args = ["fetch", target, f"+{CHECKPOINT_BRANCH}:{CHECKPOINT_BRANCH}"]
-                    max_fetch_attempts = 3
-                    fetch_timeout = 45
-                    for attempt in range(1, max_fetch_attempts + 1):
-                        try:
-                            self._run_git(
-                                repo_path,
-                                fetch_args,
-                                timeout=fetch_timeout,
-                                github_token=github_token,
-                            )
-                            break
-                        except (
-                            subprocess.TimeoutExpired,
-                            CheckpointError,
-                        ) as exc:
-                            if attempt < max_fetch_attempts:
-                                backoff = 2**attempt
-                                logger.warning(
-                                    "Checkpoint fetch failed, retrying",
-                                    attempt=attempt,
-                                    max_attempts=max_fetch_attempts,
-                                    backoff_seconds=backoff,
-                                    checkpoint_id=checkpoint.id,
-                                    error=str(exc),
-                                )
-                                time.sleep(backoff)
-                            else:
-                                logger.error(
-                                    "Checkpoint fetch failed after all retries",
-                                    attempts=max_fetch_attempts,
-                                    checkpoint_id=checkpoint.id,
-                                    error=str(exc),
-                                )
-                                raise
-                    self._run_git(
-                        repo_path,
-                        [
-                            "worktree",
-                            "add",
-                            "--detach",
-                            str(temp_path),
-                            CHECKPOINT_BRANCH,
-                        ],
-                    )
-                else:
-                    # Delete any stale local branch before creating orphan.
-                    # The branch may exist locally from a different remote
-                    # (e.g., checkpoints were previously stored in the source
-                    # repo before migrating to an external checkpoint repo).
-                    self._run_git(
-                        repo_path,
-                        ["branch", "-D", CHECKPOINT_BRANCH],
-                        check=False,
-                    )
-                    self._run_git(
-                        repo_path,
-                        ["worktree", "add", "--detach", str(temp_path)],
-                    )
-                    self._run_git(
-                        str(temp_path),
-                        ["checkout", "--orphan", CHECKPOINT_BRANCH],
-                    )
-                    self._run_git(str(temp_path), ["rm", "-rf", "."], check=False)
-
+                # Wrap the worktree-add and the checkpoint work so that
+                # `worktree remove` and `worktree prune` always run, even
+                # if `worktree add` itself raises (which is exactly the
+                # failure mode #2069 reports). Without this, a partially-
+                # created `.git/worktrees/<basename>` entry would persist
+                # until the next successful store's cleanup pruned it.
                 try:
+                    if branch_exists:
+                        # Force-update the local branch to match the remote.
+                        # The + prefix handles the case where the local branch
+                        # has diverged (e.g., from a different remote or
+                        # concurrent checkpoint pushes).
+                        # Retry with backoff: network fetches can be slow or
+                        # transiently fail, especially during container shutdown.
+                        fetch_args = [
+                            "fetch",
+                            target,
+                            f"+{CHECKPOINT_BRANCH}:{CHECKPOINT_BRANCH}",
+                        ]
+                        max_fetch_attempts = 3
+                        fetch_timeout = 45
+                        for attempt in range(1, max_fetch_attempts + 1):
+                            try:
+                                self._run_git(
+                                    repo_path,
+                                    fetch_args,
+                                    timeout=fetch_timeout,
+                                    github_token=github_token,
+                                )
+                                break
+                            except (
+                                subprocess.TimeoutExpired,
+                                CheckpointError,
+                            ) as exc:
+                                if attempt < max_fetch_attempts:
+                                    backoff = 2**attempt
+                                    logger.warning(
+                                        "Checkpoint fetch failed, retrying",
+                                        attempt=attempt,
+                                        max_attempts=max_fetch_attempts,
+                                        backoff_seconds=backoff,
+                                        checkpoint_id=checkpoint.id,
+                                        error=str(exc),
+                                    )
+                                    time.sleep(backoff)
+                                else:
+                                    logger.error(
+                                        "Checkpoint fetch failed after all retries",
+                                        attempts=max_fetch_attempts,
+                                        checkpoint_id=checkpoint.id,
+                                        error=str(exc),
+                                    )
+                                    raise
+                        self._run_git(
+                            repo_path,
+                            [
+                                "worktree",
+                                "add",
+                                "--detach",
+                                str(temp_path),
+                                CHECKPOINT_BRANCH,
+                            ],
+                        )
+                    else:
+                        # Delete any stale local branch before creating orphan.
+                        # The branch may exist locally from a different remote
+                        # (e.g., checkpoints were previously stored in the source
+                        # repo before migrating to an external checkpoint repo).
+                        self._run_git(
+                            repo_path,
+                            ["branch", "-D", CHECKPOINT_BRANCH],
+                            check=False,
+                        )
+                        self._run_git(
+                            repo_path,
+                            ["worktree", "add", "--detach", str(temp_path)],
+                        )
+                        self._run_git(
+                            str(temp_path),
+                            ["checkout", "--orphan", CHECKPOINT_BRANCH],
+                        )
+                        self._run_git(str(temp_path), ["rm", "-rf", "."], check=False)
+
                     checkpoint_path = get_checkpoint_path(temp_path / "checkpoints", checkpoint.id)
 
                     save_checkpoint_v2(checkpoint, checkpoint_path)
@@ -1071,7 +1081,8 @@ class CheckpointHandler:
                         check=False,
                     )
                     # Drop any stale .git/worktrees/<basename> entry so a
-                    # failed remove doesn't leak across attempts.
+                    # failed remove (or a failed worktree add) doesn't
+                    # leak across attempts.
                     self._run_git(
                         repo_path,
                         ["worktree", "prune"],
@@ -1313,21 +1324,25 @@ class CheckpointHandler:
 _checkpoint_handler: CheckpointHandler | None = None
 _handler_lock = threading.Lock()
 
-# Per-(repo_path, target) locks serializing concurrent checkpoint stores.
+# Per-repo_path locks serializing concurrent checkpoint stores.
 # Concurrent threads operating on the same source repo race on
-# .git/worktrees state and ref locks, producing 'worktree add' failures
-# and stale worktree directories. See #2069.
-_repo_locks: dict[tuple[str, str], threading.Lock] = {}
+# .git/worktrees state and ref locks under repo_path/.git/, producing
+# 'worktree add' failures and stale worktree directories. The lock is
+# keyed by repo_path only — the target (origin vs. external checkpoint
+# repo) does not affect the local .git state being contended.
+# Not trimmed: the set of distinct repo_paths in a gateway process is
+# small and stable in practice (one entry per source repo seen).
+# See #2069.
+_repo_locks: dict[str, threading.Lock] = {}
 _repo_locks_guard = threading.Lock()
 
 
-def _get_repo_lock(repo_path: str, target: str) -> threading.Lock:
-    key = (repo_path, target)
+def _get_repo_lock(repo_path: str) -> threading.Lock:
     with _repo_locks_guard:
-        lock = _repo_locks.get(key)
+        lock = _repo_locks.get(repo_path)
         if lock is None:
             lock = threading.Lock()
-            _repo_locks[key] = lock
+            _repo_locks[repo_path] = lock
         return lock
 
 

--- a/gateway/tests/test_checkpoint_handler.py
+++ b/gateway/tests/test_checkpoint_handler.py
@@ -642,6 +642,155 @@ class TestStoreCheckpointV2GitOps:
         assert branch_delete_calls[0][2].get("check") is False
 
 
+class TestStoreCheckpointV2Concurrency:
+    """Tests for per-repo serialization of store_checkpoint_v2 (issue #2069)."""
+
+    def _make_checkpoint(self):
+        from egg_contracts.checkpoints import (
+            CheckpointV2,
+            SessionMetadata,
+            TriggerType,
+        )
+
+        now = datetime.now(UTC)
+        return CheckpointV2(
+            id="ckpt-a1b2c3d4e5f67890",
+            trigger_type=TriggerType.COMMIT,
+            session_id="test-container",
+            commit_sha="abc123def456789012345678901234567890abcd",
+            session=SessionMetadata(session_id="test-container", started_at=now),
+            created_at=now,
+            session_started_at=now,
+        )
+
+    def test_worktree_prune_called_in_cleanup(self):
+        """Cleanup runs `git worktree prune` to drop stale worktree entries."""
+        import checkpoint_handler
+
+        handler = checkpoint_handler.CheckpointHandler(github_token="test-token")
+
+        git_calls = []
+
+        def track_run_git(cwd, args, **kwargs):
+            git_calls.append((cwd, args, kwargs))
+            return MagicMock(returncode=0, stdout="", stderr="")
+
+        handler._run_git = track_run_git
+        handler._branch_exists = MagicMock(return_value=True)
+
+        try:
+            handler.store_checkpoint_v2(self._make_checkpoint(), "/fake/repo")
+        except Exception:
+            pass
+
+        prune_calls = [c for c in git_calls if c[1][:2] == ["worktree", "prune"]]
+        assert len(prune_calls) == 1, "Expected exactly one `worktree prune` call"
+        assert prune_calls[0][0] == "/fake/repo"
+        assert prune_calls[0][2].get("check") is False
+
+    def test_concurrent_stores_on_same_repo_serialized(self):
+        """Two threads storing into the same (repo, target) run sequentially."""
+        import threading
+        import time
+
+        import checkpoint_handler
+
+        # Reset the per-repo lock dict so this test is isolated.
+        checkpoint_handler._repo_locks.clear()
+
+        handler = checkpoint_handler.CheckpointHandler(github_token="test-token")
+
+        in_flight = 0
+        max_in_flight = 0
+        observe_lock = threading.Lock()
+
+        def track_run_git(cwd, args, **kwargs):
+            nonlocal in_flight, max_in_flight
+            with observe_lock:
+                in_flight += 1
+                max_in_flight = max(max_in_flight, in_flight)
+            time.sleep(0.05)
+            with observe_lock:
+                in_flight -= 1
+            return MagicMock(returncode=0, stdout="", stderr="")
+
+        handler._run_git = track_run_git
+        handler._branch_exists = MagicMock(return_value=True)
+
+        def run_store():
+            try:
+                handler.store_checkpoint_v2(
+                    self._make_checkpoint(),
+                    "/fake/repo",
+                    checkpoint_repo="owner/repo-checkpoints",
+                )
+            except Exception:
+                pass
+
+        threads = [threading.Thread(target=run_store) for _ in range(3)]
+        for t in threads:
+            t.start()
+        for t in threads:
+            t.join()
+
+        # If serialization works, only one thread is ever inside _run_git
+        # at a time for the same (repo_path, target) key.
+        assert max_in_flight == 1, f"Expected serialized git calls, saw {max_in_flight} concurrent"
+
+    def test_concurrent_stores_on_different_repos_not_serialized(self):
+        """Different (repo, target) keys take different locks and run in parallel."""
+        import threading
+        import time
+
+        import checkpoint_handler
+
+        checkpoint_handler._repo_locks.clear()
+
+        handler = checkpoint_handler.CheckpointHandler(github_token="test-token")
+
+        in_flight = 0
+        max_in_flight = 0
+        observe_lock = threading.Lock()
+        barrier = threading.Barrier(2)
+
+        def track_run_git(cwd, args, **kwargs):
+            nonlocal in_flight, max_in_flight
+            with observe_lock:
+                in_flight += 1
+                max_in_flight = max(max_in_flight, in_flight)
+            # Synchronize so both threads are guaranteed to overlap if
+            # the locks are independent.
+            try:
+                barrier.wait(timeout=2.0)
+            except threading.BrokenBarrierError:
+                pass
+            time.sleep(0.05)
+            with observe_lock:
+                in_flight -= 1
+            return MagicMock(returncode=0, stdout="", stderr="")
+
+        handler._run_git = track_run_git
+        handler._branch_exists = MagicMock(return_value=True)
+
+        def run_store(repo_path):
+            try:
+                handler.store_checkpoint_v2(self._make_checkpoint(), repo_path)
+            except Exception:
+                pass
+
+        t1 = threading.Thread(target=run_store, args=("/fake/repo-a",))
+        t2 = threading.Thread(target=run_store, args=("/fake/repo-b",))
+        t1.start()
+        t2.start()
+        t1.join()
+        t2.join()
+
+        assert max_in_flight >= 2, (
+            "Expected concurrent execution across different repos, "
+            f"saw max_in_flight={max_in_flight}"
+        )
+
+
 class TestStoreCheckpointV2RemoteTarget:
     """Tests for store_checkpoint_v2 remote URL resolution (issue #1767).
 

--- a/gateway/tests/test_checkpoint_handler.py
+++ b/gateway/tests/test_checkpoint_handler.py
@@ -667,6 +667,9 @@ class TestStoreCheckpointV2Concurrency:
         """Cleanup runs `git worktree prune` to drop stale worktree entries."""
         import checkpoint_handler
 
+        # Reset the per-repo lock dict so this test is isolated.
+        checkpoint_handler._repo_locks.clear()
+
         handler = checkpoint_handler.CheckpointHandler(github_token="test-token")
 
         git_calls = []
@@ -688,8 +691,38 @@ class TestStoreCheckpointV2Concurrency:
         assert prune_calls[0][0] == "/fake/repo"
         assert prune_calls[0][2].get("check") is False
 
+    def test_worktree_prune_runs_when_worktree_add_fails(self):
+        """If `worktree add` itself raises, cleanup still runs `worktree prune`."""
+        import checkpoint_handler
+
+        checkpoint_handler._repo_locks.clear()
+
+        handler = checkpoint_handler.CheckpointHandler(github_token="test-token")
+
+        git_calls = []
+
+        def track_run_git(cwd, args, **kwargs):
+            git_calls.append((cwd, args, kwargs))
+            if args[:2] == ["worktree", "add"]:
+                raise checkpoint_handler.CheckpointError("simulated worktree add failure")
+            return MagicMock(returncode=0, stdout="", stderr="")
+
+        handler._run_git = track_run_git
+        handler._branch_exists = MagicMock(return_value=True)
+
+        # store_checkpoint_v2 swallows exceptions and returns False; we just
+        # need it to complete and run the finally block.
+        result = handler.store_checkpoint_v2(self._make_checkpoint(), "/fake/repo")
+        assert result is False
+
+        prune_calls = [c for c in git_calls if c[1][:2] == ["worktree", "prune"]]
+        assert len(prune_calls) == 1, (
+            "Expected `worktree prune` to run even when `worktree add` fails"
+        )
+        assert prune_calls[0][0] == "/fake/repo"
+
     def test_concurrent_stores_on_same_repo_serialized(self):
-        """Two threads storing into the same (repo, target) run sequentially."""
+        """Two threads storing into the same repo_path run sequentially."""
         import threading
         import time
 
@@ -734,11 +767,11 @@ class TestStoreCheckpointV2Concurrency:
             t.join()
 
         # If serialization works, only one thread is ever inside _run_git
-        # at a time for the same (repo_path, target) key.
+        # at a time for the same repo_path.
         assert max_in_flight == 1, f"Expected serialized git calls, saw {max_in_flight} concurrent"
 
     def test_concurrent_stores_on_different_repos_not_serialized(self):
-        """Different (repo, target) keys take different locks and run in parallel."""
+        """Different repo_paths take different locks and run in parallel."""
         import threading
         import time
 

